### PR TITLE
fix(gate-context): make --angles optional; CLI resolves angles dynamically

### LINKED ISSUE #1366: investigate: dynamic gate-angle resolution inconsistent between write-gate-context CLI and buildGateContext API
## Summary

Dynamic gate-angle resolution is inconsistent between the CLI and the programmatic API. `write-gate-context.mjs` requires `--angles` and uses the passed list **verbatim**, bypassing `resolveGateAnglesDynamic` entirely — while the programmatic `buildGateContext` API it sits next to *does* resolve dynamically from config + diff. The result: a caller running the gate context CLI gets no dynamic trimming even when `dynamicAngles: true` and a `--base` diff are both present.

## Scope and context

Surfaced while running the `pre_approval_gate` for #1365. `write-gate-context.mjs --gate pre_approval_gate --base origin/main` was invoked with all 17 configured `preApproval` angles passed via `--angles`. The resulting artifact carried `resolvedAngles` = all 17 — none trimmed — despite `.devloops` having `gates.preApproval.dynamicAngles: true` and a resolvable `--base` diff.

Investigation found two coupled gaps in `scripts/github/write-gate-context.mjs`:

1. **`--angles` is required.** `main()` rejects the call without it (the missing-arg guard treats `angles` as required).
2. **The CLI uses `--angles` verbatim.** `main()` calls the low-level `writeGateContext(options)` directly, which stamps `resolvedAngles: [...options.angles]`. It never calls `resolveGateAnglesDynamic`, so the configured `dynamicAngles` flag has no effect on the CLI path.

Meanwhile the programmatic `buildGateContext` API (same file) resolves angles itself:

```js
const resolverResult = await resolveGateAnglesDynamic(input.config, configKey, { diff: input.diff });
```

`resolveGateAnglesDynamic` (in `packages/core/src/config/config.mjs`) returns the full static pool whenever `dynamicAngles` is off **or** no diff is given; with both on, it trims to the mandatory floor + diff-selected candidates. So the trimming logic exists and works — the CLI simply doesn't reach it.

### Inconsistencies to resolve

- The CLI requires `--angles`; the API does not. `--angles` should be optional, defaulting to dynamic resolution from the loaded config + captured `--base` diff.
- When `--angles` IS passed, decide and document whether it is an override (verbatim) or still flows through the resolver. Today it is silently verbatim, which defeats `dynamicAngles`.
- The CLI does not load dev-loop config today; dynamic resolution on the CLI path requires loading config (or reusing the `buildGateContext` path) — confirm the sanctioned config-loading entrypoint.
- Verify the CLI and API paths produce identical angle sets for the same config + diff, and that a contract test pins this (so the CLI can't drift from the API again).
- Confirm why the API path, when given a diff, trims correctly and exercise it end-to-end (the investigation should prove the trim works, not assume it).

## Acceptance criteria

- [x] `write-gate-context.mjs` runs successfully **without** `--angles`, resolving angles dynamically from config + the `--base` diff.
- [x] `--angles`, when supplied, has documented semantics (override vs. resolver-fed) and behaves accordingly.
- [x] The CLI path and the `buildGateContext` API produce the same `resolvedAngles` for the same config + diff.
- [x] A contract/unit test pins the CLI dynamic-resolution behavior (angles trim when `dynamicAngles: true` + diff present; fall back to the static pool otherwise).
- [x] The `pre_approval_gate` context build for a small PR no longer fans out all configured angles when dynamic resolution would trim them.

## Definition of done

- [x] Inconsistency resolved in source (CLI resolves dynamically by default; `--angles` optional).
- [x] Tests added/updated and green; `npm run verify` passes.
- [ ] Merged through the full loop (gates, Copilot, green CI).

## Non-goals

- Changing which angles are mandatory or the configured `preApproval` pool itself.
- Rewriting the diff analyzer / change classifier — only wiring the existing resolver into the CLI path.
- Re-evaluating the fan-out parallelism cap.

## Validation

- `node scripts/github/write-gate-context.mjs --repo mfittko/dev-loops --pr <n> --gate pre_approval_gate --head-sha <sha> --base origin/main` (no `--angles`) resolves a trimmed angle set.
- `npm run verify` green.

## Context

Discovered during the #1365 `pre_approval_gate` run: the gate context kept all 17 `preApproval` angles because the CLI bypassed `resolveGateAnglesDynamic`. Filing as investigation so the dynamic-angles feature is made consistent end-to-end rather than worked around per-call.

### PR DESCRIPTION
**Pull request by mfittko**
## Summary

The `write-gate-context.mjs` CLI required `--angles` and used the passed list verbatim, bypassing `resolveGateAnglesDynamic` entirely — while the sibling programmatic `buildGateContext` API resolves dynamically from config + diff. So a caller running the CLI got no dynamic trimming even with `dynamicAngles: true` and a `--base` diff (all configured angles were kept). This makes `--angles` optional and routes the CLI through the same dynamic resolver the API uses.

## Scope and context

- `scripts/github/write-gate-context.mjs` `main()`: when `--angles` is omitted, load config via `loadDevLoopConfig` and resolve angles through `resolveGateAnglesDynamic(config, configKey, { diff })` — the same call `buildGateContext` makes. The `--base` diff is captured before resolution so the resolver can trim. When `--angles` IS supplied, it stays a verbatim override (documented escape hatch; dynamic resolution bypassed).
- Removed `angles` from the required-arg missing-check; updated USAGE text.
- Added stderr warnings on the no-`--angles` path (warn-and-proceed, non-fatal): unvalidatable `.devloops` config (proceeds on the merged fallback config), zero angles resolved for the gate (so a hollow context bundle is never written silently), and a caller-supplied `--rationale` discarded because angles were resolved dynamically.
- Tests added for the new dynamic-resolution behaviors and their edge cases (optionality, docs-only trimming, CLI/API parity, the verbatim override, the static-pool fallback, the warn-and-proceed paths, and resolver-derived rationale) — see `test/github/write-gate-context.test.mjs`.

Filed as investigation #1366 — discovered during the #1365 `pre_approval_gate` run, where passing all 17 `preApproval` angles kept all 17 (none trimmed).

## Acceptance criteria

- [x] `write-gate-context.mjs` runs successfully without `--angles`, resolving dynamically from config + the `--base` diff.
- [x] `--angles`, when supplied, has documented semantics (verbatim override) and behaves accordingly.
- [x] The CLI path and the `buildGateContext` API produce the same `resolvedAngles` for the same config + diff.
- [x] A contract/unit test pins the CLI dynamic-resolution behavior (trims when `dynamicAngles: true` + diff; falls back to the static pool otherwise).
- [x] The `pre_approval_gate` context build for a small PR no longer fans out all configured angles when dynamic resolution would trim them.

## Definition of done

- [x] Source fixed (CLI resolves dynamically by default; `--angles` optional).
- [x] Tests added/updated and green; `npm run verify` passes.
- [ ] Merged through the full loop (draft_gate → pre_approval_gate → Copilot → green CI).

## Non-goals

- Changing which angles are mandatory or the configured `preApproval` pool itself.
- Rewriting the diff analyzer / change classifier — only wiring the existing resolver into the CLI path.
- Re-evaluating the fan-out parallelism cap.

## Validation

- `node scripts/github/write-gate-context.mjs --repo mfittko/dev-loops --pr 1366 --gate pre_approval_gate --head-sha <sha> --base origin/main` (no `--angles`) resolves a trimmed set: **17 → 5 angles** (`pr-checklist-matrix`, `acceptance-criteria`, `yagni`, `contradiction-lens`, `renderer-security`).
- `npm run verify` green; `test/github/write-gate-context.test.mjs` passing.

Closes #1366

### CODE REVIEW

**Comment by mfittko** [comment]
### Gate review: `draft_gate`

**Reviewed head SHA:** `494aac68c05e564cb1e6e2aa447248f555983437`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 8 angles reviewed; 3 findings (see per-angle breakdown below).

- `pr-description` → clean
- `scope` → clean
- `coverage` → findings_present
  - [defer] "CLI --rationale supplied WITHOUT --angles is ignored" test exercises the options.rationale.length > 0 branch (scripts/github/write-gate-context.mjs:1145-1148) and verifies the functional outcome (resolver rationale persisted, stale rationale discarded), but unlike the sibling malformed-config and zero-resolved-angles tests it does not capture stderr and assert the specific discard warning text ("--rationale was supplied without --angles"). The branch is covered; only the exact warning message is unverified. (`test/github/write-gate-context.test.mjs:1030`)
  - [defer] All 8 new CLI dynamic-angle-resolution tests use --gate draft_gate; none exercise the same new main() block with --gate pre_approval_gate (configKey "preApproval"). mapGateToConfigKey itself is already unit-tested for both gates and buildGateContext already proves the preApproval mapping end-to-end, so this is a purely combinatorial gap (gate-name x new-CLI-path), not an untested code path. (`test/github/write-gate-context.test.mjs:999`)
- `correctness` → clean
- `contract-surface` → clean
- `input-validation` → findings_present
  - [defer] The new zero-resolved-angles warning path (main(), line ~1135) is fed by rationaleFromResolver's null->[] coercion (line 69-71): resolveGateAngles returns recommendedAngles=null for a gate with NO configured `angles` and NO `mandatoryAngles` at all (an 'unrestricted/unconfigured' signal), which is a different condition than a gate explicitly configured with `angles: []`/`mandatoryAngles: []` (a deliberate zero-angle pool, exercised by the new test). Both collapse to resolvedAngles=[] and the SAME warning text ('Check the gate's configured angles/mandatoryAngles'), which is misleading for the null case (there is no gate config section to check) and silently discards the distinct 'unrestricted' semantic that resolveGateAngles otherwise carries. Pre-existing in…[truncated 347 chars]

**Next action:** Mark PR ready for review, then request Copilot review.

File: test/github/write-gate-context.test.mjs [hunk 1]
```diff
@@ -91,6 +91,49 @@ const DOCS_ONLY_DIFF = {
...
+    angles: ["scope", "coverage", "correctness", "docs", "link-check", "config-drift"],
+    mandatoryAngles: ["gate-evidence"],
+    excludeAngles: [],
+    dynamicAngles: true,
+    ...overrides,
+  };
+  const lines = ["version: 1", "gates:", "  draft:"];
+  lines.push(`    dynamicAngles: ${draft.dynamicAngles}`);
+  lines.push(`    excludeAngles: []`);
+  lines.push("    angles:");
```
[the remark this answers is not in the document]
Comment (mfittko) [reply test/github/write-gate-context.test.mjs:130]: Fixed in ce82eb58736e4ddb5c2308a4f2b630371b18d038. writeDraftDevLoops() now emits excludeAngles from the merged draft object (`excludeAngles: ${JSON.stringify(draft.excludeAngles)}`) instead of a hard-coded `[]`, so an overrides.excludeAngles reaches the loaded .devloops config and the helper truly mirrors draftConfig(). Added a regression test that sets excludeAngles: ["coverage"] and asserts the resolved angle set excludes it.

File: scripts/github/write-gate-context.mjs [hunk 2]
```diff
@@ -1104,6 +1105,50 @@ export async function main(argv = process.argv.slice(2), { repoRoot = process.cw
...
+      }
+      options.angles = resolvedAngles;
+      // Resolver-derived rationale is authoritative here: a caller cannot supply
+      // meaningful rationale for angles it did not name (angles were just
+      // resolved dynamically above), so any --rationale the caller passed is
+      // ignored rather than persisted as a stale mismatch.
+      if (options.rationale.length > 0) {
+        process.stderr.write(
+          "[write-gate-context] warning: --rationale was supplied without --angles; ignoring it in favor of the dynamically-resolved rationale.\n",
+        );
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/write-gate-context.mjs:1148]: Fixed in 70109d7f93b191d48c7d95c50c29d619c4aba360. The warning no longer implies dynamic trimming — it now reads "--rationale was supplied without --angles; ignoring it in favor of the resolver-derived rationale (angles were resolved from config rather than supplied via --angles)".

File: scripts/github/write-gate-context.mjs [hunk 3]
```diff
@@ -1104,6 +1105,50 @@ export async function main(argv = process.argv.slice(2), { repoRoot = process.cw
...
+          `[write-gate-context] warning: dev-loop config could not be fully loaded/validated; resolving angles from the merged fallback config. errors=${JSON.stringify(configErrors)}\n`,
+        );
+      }
+      const configKey = mapGateToConfigKey(options.gate);
+      const resolverResult = await resolveGateAnglesDynamic(config, configKey, { diff });
+      const { resolvedAngles, rationale } = rationaleFromResolver(resolverResult);
+      if (resolvedAngles.length === 0) {
+        process.stderr.write(
+          `[write-gate-context] warning: dynamic angle resolution produced zero angles for gate ${options.gate}; the gate-context bundle carries no review angles. Check the gate's configured angles/mandatoryAngles.\n`,
+        );
```
[the remark this answers is not in the document]
Comment (mfittko) [reply scripts/github/write-gate-context.mjs:1138]: Fixed in 70109d7f93b191d48c7d95c50c29d619c4aba360. The warning no longer says "dynamic angle resolution" — it now reads "angle resolution produced zero angles for gate <gate>...", since this path also runs in static-fallback mode (dynamicAngles:false or no diff).

**Comment by mfittko** [comment]
### Gate review: `pre_approval_gate`

**Reviewed head SHA:** `70109d7f93b191d48c7d95c50c29d619c4aba360`
**Verdict:** clean
**Execution mode:** fanout_fanin

**Findings summary:** 5 angles reviewed; 1 finding (see per-angle breakdown below).; Copilot review rounds exhausted (2/2); current head has zero unresolved threads and green or credibly green CI, so pre_approval_gate fallback is allowed without another Copilot re-request.

- `pr-checklist-matrix` → clean
- `acceptance-criteria` → clean
- `yagni` → clean
- `contradiction-lens` → findings_present
  - [defer] Assertion comment claims resolvedAngles===[] here is 'matching the resolver's null->[] mapping' (i.e. rationaleFromResolver's `Array.isArray(resolverResult?.recommendedAngles) ? ... : []` null-fallback branch). It isn't: this fixture has angles:[] mandatoryAngles:[] excludeAngles:[] with dynamicAngles:false, so resolveGateAngles returns [] (not null, since gateConfig.angles is [] not null) and resolveGateAnglesDynamic's dynamicAngles-off branch returns recommendedAngles:[] directly. Verified empirically: resolveGateAnglesDynamic(config, 'draft', {diff:null}) returns {"recommendedAngles":[],...} — never null. rationaleFromResolver's Array.isArray check takes the true branch (already-array), not the null-coalescing fallback the comment describes. The test's outcome is correct; the comment misattributes the mechanism. (`test/github/write-gate-context.test.mjs:1292`)
- `renderer-security` → clean

**Next action:** Stop at human approval checkpoint; PR is merge-ready pending authorization. Copilot converged at the round cap (2 rounds; all threads addressed and resolved), so no further Copilot re-request was issued per the round-exhaustion fallback.
